compiler/src/dotty/tools/dotc/transform/UnusedRefs.scala

```diff
@@ -0,0 +1,58 @@
+package dotty.tools.dotc.transform
+
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.ast.Trees._
+import dotty.tools.dotc.core.Contexts._
+import dotty.tools.dotc.core.Flags._
+import dotty.tools.dotc.core.Phases
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.core.Types._
+import dotty.tools.dotc.transform.MegaPhase.MiniPhase
+
+/** This phase removes all references and calls to unused methods or vals
+ *
+ *   if `unused def f(x1,...,xn): T = ...`
+ *   then `f(y1,...,yn)` --> `y1; ...; yn; (default value for T)`
```

Comment:
Why not use `???` instead of default value? If something goes wrong later you will fail loudly instead of silently, which is better.

Reply:
These default values are intended to stay there, it is the simplest tree the can be both used when `f(...)` is a statement or an expression. These are trivially eliminated by the optimizer.
This way `UnusedRefs` does not need to know and special case unused function applications in different positions. The same applies to member selection and identifiers.